src/Mvc/Mvc.Core/src/Routing/DynamicControllerEndpointMatcherPolicy.cs

```diff
@@ -83,30 +89,56 @@ public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
                 }
 
                 var endpoint = candidates[i].Endpoint;
+                var originalValues = candidates[i].Values;
+
+                RouteValueDictionary dynamicValues = null;
+
+                // We don't expext both of these to be provided, and they are internal so there's
+                // no realistic way this could happen.
+                var dynamicControllerMetadata = endpoint.Metadata.GetMetadata<DynamicControllerMetadata>();
+                var transformerMetadata = endpoint.Metadata.GetMetadata<DynamicControllerRouteValueTransformerMetadata>();
+                if (dynamicControllerMetadata != null)
+                {
+                    dynamicValues = dynamicControllerMetadata.Values;
+                }
+                else if (transformerMetadata != null)
+                {
+                    var transformer = (DynamicRouteValueTransformer)httpContext.RequestServices.GetRequiredService(transformerMetadata.SelectorType);
+                    dynamicValues = await transformer.TransformAsync(httpContext, originalValues);
+                }
+                else
+                {
+                    // Not a dynamic controller.
+                    continue;
+                }
 
-                var metadata = endpoint.Metadata.GetMetadata<DynamicControllerMetadata>();
-                if (metadata == null)
+                if (dynamicValues == null)
                 {
+                    candidates.ReplaceEndpoint(i, null, null);
```

Comment:
What does this do?

Reply:
It "nulls out" the match, in one go it sets the endpoint and rvd to null, and marks it as a non-match.